scripts/maintenance/migrate_project/src/db.py

```diff
@@ -0,0 +1,163 @@
+    project_id = project["uuid"]
+
+    if _project_uuid_exists_in_destination(dst_conn, project_id):
+        print(f"\n\n>>>>>>>>>SKIPPING, sync for {project=}\n\n")
+        skipped_projects.append(project)
+        raise Exception("could not continue, main project already exists on remote!")
+
+    projects_to_migrate.append(project)
+
+    if hidden_projects_for_user:
+        # extract all hidden projects and check if they require syncing
+        hidden_projects_cursor = _get_hidden_project(src_conn, hidden_projects_for_user)
+        for hidden_result in hidden_projects_cursor:
+            hidden_project = dict(hidden_result.items())
+            if _project_uuid_exists_in_destination(dst_conn, hidden_project["uuid"]):
+                print(f"\n\n>>>>>>>>>SKIPPING, sync for {project=}\n\n")
+                skipped_projects.append(project)
+                continue
+
+            projects_to_migrate.append(hidden_project)
+
+    # check file_meta_data in the projects to migrate
+    for project in projects_to_migrate:
+        user_id = project["prj_owner"]
+        project_id = project["uuid"]
+
+        files_metadata_cursor = _get_file_meta_data_without_soft_links(
+            connection=src_conn, user_id=user_id, project_id=project_id
+        )
+        for result in files_metadata_cursor:
+            file_meta_data = dict(result.items())
+            file_uuid = file_meta_data["file_uuid"]
+
+            if _meta_data_exists_in_destination(dst_conn, file_uuid):
+                print(f"\n\n>>>>>>>>>SKIPPING, sync for {file_meta_data=}\n\n")
+                skipped_files_meta_data.append(file_meta_data)
+                continue
+
+            files_meta_data_to_migrate.append(file_meta_data)
+
+    if len(skipped_projects) > 0:
+        print("SKIPPED projects", len(skipped_projects))
+    if len(skipped_files_meta_data) > 0:
+        print("SKIPPED files", len(skipped_files_meta_data))
+
+    print("Projects to move", len(projects_to_migrate))
+    print("Files to move", len(files_meta_data_to_migrate))
+    print("Projects to SKIP", len(skipped_projects))
+    print("Files to SKIP", len(skipped_files_meta_data))
+
+    # if files and projects already exist
+    if len(skipped_files_meta_data) > 0 or len(skipped_projects) > 0:
+        print(
+            "Projects skipped uuid(primary keys) listing:",
+            [x["uuid"] for x in skipped_projects],
+        )
+        print(
+            "File meta data skipped file_uuid(primary keys) listing:",
+            [x["file_uuid"] for x in skipped_files_meta_data],
+        )
+        raise Exception(
+            "Could not continue migration, some projects or files already exist."
+        )
+
+    return projects_to_migrate, files_meta_data_to_migrate
+
+
+def insert_file_meta_data(connection: Connection, data: Dict[str, Any]) -> None:
+    insert_stmt = insert(file_meta_data).values(**data)
+    on_update_stmt = insert_stmt.on_conflict_do_update(
```

Comment:
If in between checking for the UUID already existing (and giving an error) and this statement by unlikely unluck a file_meta_data entry with the UUID was created by an unrealted user using osparc, his/her data would be updated/overwritten. However, this scenario is highly unlikely
I think it could still be safer to not update on conflict, but throw.

Reply:
Good point. See my answer below.